src/storage/store/column_reader_writer.cpp

```diff
@@ -186,10 +186,18 @@ class DefaultColumnReadWriter : public ColumnReadWriter {
             KU_ASSERT(isPageIdxValid(pageCursor.pageIdx, chunkMeta));
             if (!filterFunc.has_value() ||
                 filterFunc.value()(numValuesScanned, numValuesScanned + numValuesToScanInPage)) {
-                readFromPage(transaction, pageCursor.pageIdx, [&](uint8_t* frame) -> void {
-                    readFunc(frame, pageCursor, result, numValuesScanned + startOffsetInResult,
-                        numValuesToScanInPage, chunkMeta.compMeta);
-                });
+
+                // to keep the size of the lambda small
+                // we pack all the arguments to readFunc into a tuple and capture the tuple by
+                // this makes it more likely that we avoid any heap allocations when the lambda is
+                // copied into an std::function
+                const auto startReadOffset = numValuesScanned + startOffsetInResult;
+                auto readFuncArgs = std::tie(pageCursor, result, startReadOffset,
+                    numValuesToScanInPage, chunkMeta.compMeta);
+                readFromPage(transaction, pageCursor.pageIdx,
+                    [&readFunc, &readFuncArgs](uint8_t* frame) -> void {
+                        std::apply(readFunc, std::tuple_cat(std::make_tuple(frame), readFuncArgs));
+                    });
```

Comment:
```suggestion
                // to keep the size of the lambda small
                // we pack all the arguments to readFunc into a tuple and capture the tuple by
                // this makes it more likely for the std::function to be stack-allocated
                const auto startReadOffset = numValuesScanned + startOffsetInResult;
                auto readFuncArgs = std::tie(pageCursor, result, startReadOffset,
                    numValuesToScanInPage, chunkMeta.compMeta);
                readFromPage(transaction, pageCursor.pageIdx,
                    [&readFunc, &readFuncArgs](uint8_t* frame) -> void {
                        std::apply(readFunc, std::tuple_cat(std::make_tuple(frame), readFuncArgs));
                    });
```

Reply:
I'm just looking into this now you mention this, but one option I ran across is that I think you should be able to get a very similar result by creating the lambda first as a variable (on the stack) and then passing it to readFromPage via `readFromPage(..., std::cref(lambdaFunc))`. That way you can still have it use automatic capture by reference without any heap allocations.

Reply:
Cool, let me try this out